### PR TITLE
feat: collapsible side navigation MAASENG-1282

### PR DESCRIPTION
**Pull request by petermakowski**
## Done

- Implement collapsible side navigation
- restructure the navigation to use vanilla classes
- highlight parent when any of its children are active
- remove hidden items when collapsed
- reorganise `l-navigation--` color modifier classes and rename to `is-maas-`
  - use `is-maas-` color className in `ThemedRadioButton`
- add global `[`  keyboard shortcut for expanding and collapsing the navigation

this will be followed up by https://warthogs.atlassian.net/browse/MAASENG-1356 to look into allowing for collapsibility across all breakpoints

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps
small screen size
- Verify that you can open and close the navigation and all items visible

medium screen size:
- Check that the navigation expands on hover
- Verify you can pin and unpin the navigation on medium screen size
- Verify that the navigation is displayed correctly and all items visible

Large screen size
- Verify that the navigation is displayed correctly and all items visible

## Fixes

- https://warthogs.atlassian.net/browse/MAASENG-1331
- https://warthogs.atlassian.net/browse/MAASENG-1282 

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

### After
Kapture 2023-02-13 at 14 24 39

### CODE REVIEW

**Comment by ndv99** [comment]
QA looks good! Just make sure to match the styling with the wireframe (font colour, all-caps on group titles, <hr /> beneath user preferences button and no <hr /> between main groups)

**Comment by amylily1011** [comment]
Wonderful! Love it! Can we remove the pin by any chance, sorry I didn't know I left a pin in the final design 😬

**Comment by amylily1011** [comment]
One more quick comment, when you collapse the nav, the highlighted tab should be on the icon. The icons will stack up closely and the selection highlight would be on the icon of that header.

**Comment by petermakowski** [comment]
> 


> Wonderful! Love it! Can we remove the pin by any chance, sorry I didn't know I left a pin in the final design 😬
Also, the collapsible icon should be on the bottom, it looks like a side chevron. So there should be no x icon on top.

Sure, I was mostly following vanilla implementation example as the design lacked some details on small/medium breakpoints. Will update.

**Comment by ndv99** [comment]
I like the new button! 

IIRC, I think the nav is meant to be collapsible at all breakpoints, so even on a large screen you should be able to collapse the nav to the side and expand it again. Judging by the wireframes this should also move the main content when the nav expands, although I think we should keep the small screen functionality you've implemented with it displaying over the content

**Comment by petermakowski** [comment]
> IIRC, I think the nav is meant to be collapsible at all breakpoints, so even on a large screen you should be able to collapse the nav to the side and expand it again. Judging by the wireframes this should also move the main content when the nav expands, although I think we should keep the small screen functionality you've implemented with it displaying over the content

Good point! To achieve this behaviour we'd need to change the approach. Overriding vanilla styles with all breakpoint conditions will get really messy. I think this warrants a separate task and a PR.

**Comment by ndv99** [comment]
## QA Review

### Fixing the logo positioning

In the screenshot from JAAS dashboard, it seems as though there is a solid background behind the logo and name, so when you scroll down, the nav items are hidden underneath that:
image

However, in this branch, if the content overflows vertically and you scroll down, you can still see the nav items beneath the logo and the name:
image

### Collapsing on small breakpoint

When the screen is shrunken horizontally, the Menu button opens the nav as expected. However, it does not close the nav. To close it, you have to click on the Collapse button. This also presents some issues - if you click on the grey highlight behind the double chevron, it toggles between "Expand" and "Collapse", without actually expanding or collapsing it. You have to click on the actual chevron for it to close.

I personally feel as though we could just keep the implementation you have for the medium breakpoint on the small one as well (doing away with the header entirely), although this is up for debate.

### Separators between last items

File: src/app/base/components/AppSideNavigation/AppSideNavCollapseToggle/AppSideNavCollapseToggle.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  setIsCollapsed,
+  className,
+}: {
+  isCollapsed: boolean;
+  setIsCollapsed: (isCollapsed: boolean) => void;
+  className?: string;
+}): JSX.Element => {
+  return (
+    <Tooltip
+      message={<>{!isCollapsed ? "collapse" : "expand"} ( [ )</>}
```
Comment (ndv99) [inline src/app/base/components/AppSideNavigation/AppSideNavCollapseToggle/AppSideNavCollapseToggle.tsx:15]: Would it be better to wrap the [ symbol in an inline code block like this? `[`

Just for greater clarity

File: src/scss/_utilities.scss [hunk 2]
```diff
@@ -220,4 +220,22 @@
...
+    "blue": #0060bf,
+    "purple": #7764d8,
+    "magenta": #974097,
+    "red": #a71b33,
+  );
+  @each $name, $color in $nav-theme-colors {
+    .is-maas-#{$name} {
+      background-color: $color !important;
+    }
+  }
```
Comment (ndv99) [inline src/scss/_utilities.scss:240]: This is really great! Really clean implementation and allows us to theme other stuff in the future as well. Nice one!

**Review by ndv99 — COMMENTED** [review]
Code looks good over all but there's a few QA issues (noted above)

**Comment by petermakowski** [comment]
> Code looks good over all but there's a few QA issues (noted above)

I should've noted that I didn't dive deeper into overwriting vanilla styles as it began feeling counter-productive as we started to diverge more and more from the original navigation spec that's implemented in vanilla.

I will update to fix it but also revisit in the ticket mention in the description to clean it up (and potentially write our own classes from scratch instead overwriting vanilla).

**Review by ndv99 — APPROVED** [review]
Looks good! 

@dgtlntv noticed a bug on the settings page with smaller breakpoints - the settings nav gets hidden behind the main nav when it becomes collapsible https://github.com/canonical/maas-ui/issues/4759

We can fix this in a follow-up PR though

**Pull request merged**